go/payments/pkg/prolly_cellwise/diff_attributor.go

```diff
@@ -0,0 +1,239 @@
+	"github.com/dolthub/dolt/go/store/prolly"
+	"github.com/dolthub/dolt/go/store/prolly/tree"
+	"github.com/dolthub/dolt/go/store/types"
+	"github.com/dolthub/dolt/go/store/val"
+	"go.uber.org/zap"
+)
+
+type diffAttributor struct {
+	tblSch           schema.Schema
+	table, prevTable *doltdb.Table
+	commitIdx        int16
+	shardMgr         *prollyShardManager
+	shard            AttributionShard
+	logger           *zap.Logger
+	format           *types.NomsBinFormat
+	kd               val.TupleDesc
+	attVd            val.TupleDesc
+
+	newData       int
+	unchangedData int
+	modifiedData  int
+	total         int
+}
+
+func newDiffAttributor(
+	ctx context.Context,
+	table, prevTable *doltdb.Table,
+	commitIdx int16,
+	shardMgr *prollyShardManager,
+	shard AttributionShard,
+	logger *zap.Logger) (*diffAttributor, error) {
+
+	if table == nil && prevTable == nil {
+		panic("|table| and |prevTable| were both nil")
+	}
+
+	var format *types.NomsBinFormat
+	var tblSch schema.Schema
+	var err error
+	if table != nil {
+		format = table.Format()
+		tblSch, err = table.GetSchema(ctx)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		format = prevTable.Format()
+		tblSch, err = prevTable.GetSchema(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	kd, attVd := getAttribDescriptorsFromTblSchema(tblSch)
+
+	return &diffAttributor{
+		tblSch:    tblSch,
+		table:     table,
+		prevTable: prevTable,
+		commitIdx: commitIdx,
+		shardMgr:  shardMgr,
+		shard:     shard,
+		logger:    logger,
+		format:    format,
+		kd:        kd,
+		attVd:     attVd,
+	}, nil
+}
+
+func (dA *diffAttributor) attributeDiffs(ctx context.Context, diffIter doltutils.ProllyDiffIter, attItr prolly.MapIter) error {
+	defer dA.incAndLog(nil)
+
+	attKey, attVal, err := attItr.Next(ctx)
+	if err != nil && err != io.EOF {
+		return err
+	}
+
+	diff, err := diffIter.Next(ctx)
+	if err != nil && err != io.EOF {
+		return err
+	}
+
+	for len(diff.Key) != 0 && len(attKey) != 0 {
+		cmp := dA.kd.Compare(val.Tuple(diff.Key), attKey)
+
+		switch {
+		case cmp < 0:
+			// Don't have an attribution for this diff (Add one)
+			err = dA.createAttrib(ctx, diff)
+			if err != nil {
+				return err
+			}
+
+			diff, err = diffIter.Next(ctx)
+			if err != nil && err != io.EOF {
+				return err
+			}
+
+		case cmp > 0:
+			// No diff for this attribution (no-op)
+			err = dA.copyAttrib(ctx, attKey, attVal)
+			if err != nil {
+				return err
+			}
+
+			attKey, attVal, err = attItr.Next(ctx)
+			if err != nil && err != io.EOF {
+				return err
+			}
+
+		case cmp == 0:
+			// A previous attribution was found for this diff
+			err = dA.updateAttrib(ctx, diff, attVal)
+			if err != nil {
+				return err
+			}
+
+			diff, err = diffIter.Next(ctx)
+			if err != nil && err != io.EOF {
+				return err
+			}
+
+			attKey, attVal, err = attItr.Next(ctx)
+			if err != nil && err != io.EOF {
+				return err
+			}
+		}
+	}
+
+	for len(diff.Key) != 0 {
+		dA.incAndLog(&dA.newData)
```

Comment:
It looks like incAndLog gets called 2x for each item.  Once here, and then again inside createAttrib.